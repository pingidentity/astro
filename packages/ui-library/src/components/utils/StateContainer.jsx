import React from "react";
import PropTypes from "prop-types";
import Utils from "../../util/Utils";

const mapNothing = () => ({});

const defaultGetPassedProps = props => props;
const containerGetPassedProps = props => props.passedProps;

export const createProgressiveState = stateDefs => ({
    passedProps,
    initialState,
    boundSetState,
    getPassedProps = defaultGetPassedProps,
}) => {
    const isControlled = name => passedProps[name] === undefined;

    // figure out which props will be controlled and create initial state for those
    const pState = stateDefs.reduce((state, { name, initial }) => {
        if (isControlled(name)) {
            return {
                ...state,
                [name]: initialState[name] === undefined ? initial : initialState[name],
                // the initialState prop should override the defined initial value
                // but only if this is a controlled prop
            };
        }
        return state;
    }, {});

    // make sure the keys provided in initialState are actually being controlled
    if (!Utils.isProduction()) {
        Object.keys(initialState).forEach(key => {
            if (pState[key] === undefined) {
                console.warn(
                    `Initial state provided for prop "${key}" that is not being controlled by the component`
                );
            }
        });
    }

    // generate custom callbacks with transform functions
    const makeTransformCallbacks = (name, callbacks = []) =>
        callbacks.reduce(
            (transformCallbacks, { name: cbName, transform, passTransformedValue = false, }) => ({
                ...transformCallbacks,
                [cbName]: (value, e) => {
                    // set the state with the transformed value
                    boundSetState((state, props) => {
                        const transformedValue = transform(value, state[name], getPassedProps(props));

                        if (passedProps[cbName]) {
                            passedProps[cbName](passTransformedValue ? transformedValue : value, e);
                        }

                        return { [name]: transformedValue };
                    });
                }
            }), {}
        );

    // get the callbacks for a single state definition
    const makeCallbacksFromDef = ({ name, setter, callbacks }) => {
        // add any custom callbacks
        const customTransformCallbacks = makeTransformCallbacks(name, callbacks);

        // return with setter if defined
        return setter ? {
            [setter]: (value, e) => {
                boundSetState({ [name]: value });
                // if a callback is provided, we'll still execute it
                if (passedProps[setter]) {
                    passedProps[setter](value, e);
                }
            },
            ...customTransformCallbacks
        } : customTransformCallbacks;
    };

    // create all the callbacks
    const pCallbacks = stateDefs.reduce((returnCallbacks, def) => {
        if (!isControlled(def.name)) {
            return returnCallbacks;
        }
        return { ...returnCallbacks, ...makeCallbacksFromDef(def) };
    }, {});

    return [pState, pCallbacks];
};

class StateContainer extends React.Component {
    static propTypes = {
        stateDefs: PropTypes.arrayOf(
            PropTypes.shape({
                name: PropTypes.string.isRequired, // name of the props
                initial: PropTypes.any.isRequired, // starting value
                setter: PropTypes.string, // name of the callback in the target component
                callbacks: PropTypes.arrayOf( // when you need something different than a simple setter
                    PropTypes.shape({
                        name: PropTypes.string,
                        // Decides whether to pass the transformed value to the callback or just the value that
                        // the base component passes.
                        passTransformedValue: PropTypes.bool,
                        transform: PropTypes.func, // this function will be applied to the value and current state
                    })
                ),
            })
        ),
        initialState: PropTypes.object,
        passedProps: PropTypes.object,
        mapToProps: PropTypes.func,
    };

    static defaultProps = {
        stateDefs: [],
        initialState: {},
        passedProps: {},
        mapToProps: mapNothing,
    };

    constructor({ stateDefs, initialState, passedProps }) {
        // the constructor has two main goals
        // 1) populate the state for any props that will be controlled
        // 2) create callback functions for those props
        super();

        const [ state, callbacks ] = createProgressiveState(stateDefs)({
            initialState, passedProps, boundSetState: this.setState.bind(this), getPassedProps: containerGetPassedProps
        });

        this.state = state;
        this.callbacks = callbacks;
    }

    render() {
        const { children, mapToProps, passedProps } = this.props;

        // alert dev if there are state/prop conflicts
        if (!Utils.isProduction()) {
            Object.keys(this.state).forEach(key => {
                if (passedProps[key] !== undefined) {
                    console.warn(
                        `The prop "${key}" is being controlled by the component because it was not initially set. ` +
                        `The newly-received value will be ignored.`
                    );
                }
            });
        }

        return children({ ...passedProps, ...this.callbacks, ...this.state, ...mapToProps(this.state, passedProps) });
    }
}

export const inStateContainer = (stateDefs, mapToProps) => WrappedComponent => ({ initialState, ...props }) => (
    <StateContainer stateDefs={stateDefs} mapToProps={mapToProps} initialState={initialState} passedProps={props}>
        {containerProps => <WrappedComponent {...containerProps} />}
    </StateContainer>
);

export const toggleTransform = (value, current) => (!current);

export default StateContainer;