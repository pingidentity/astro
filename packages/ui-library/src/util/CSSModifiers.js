export const lightInputs = "modifier_light-inputs";
export const darkInputs = "modifier_dark-inputs";
export const stripInputMargins = "modifier_strip-input-margins";

export default { lightInputs, darkInputs, stripInputMargins };
