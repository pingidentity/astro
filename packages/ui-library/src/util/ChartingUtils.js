import Color from "color";

const splitData = ({ id, data }) => data.map(point => ({ [id]: point }));

/**
 * Takes data from the format of [{ id: [data] }, { id2: [data2] }] and converts
 * it to the format of [[{ id: dataPoint }, { id2: dataPoint to }]]
 * All incoming arrays of data must be of the same length
 */
export const toRechartsDataFormat = dataSet => dataSet.reduce((acc, data) => {
    const split = splitData(data);
    return acc.length > 0
        ? acc.map((frame, index) => ({ ...frame, ...split[index] }))
        : split
    ;
}, []);

const addToExisting = (index, data, acc) => {
    const { id, data: currentData } = acc[index];
    return {
        data: [
            ...currentData,
            data,
        ],
        id
    };
};

/**
 * Takes data from the format of [[{ id: dataPoint }, { id2: dataPoint to }]] and converts
 * it to the format of [{ id: [data] }, { id2: [data2] }]
 */
export const fromRechartsDataFormat = dataSet => dataSet.reduce((acc, frame) =>
    Object.entries(frame).reduce((frameAcc, [id, data]) => {
        const idIndex = frameAcc.findIndex(entry => entry.id === id);
        return idIndex > -1
            ? [
                ...idIndex > 0 ? frameAcc.slice(0, idIndex) : [],
                addToExisting(idIndex, data, frameAcc),
                ...idIndex < frameAcc.length ? frameAcc.slice(idIndex + 1, frameAcc.length) : []
            ]
            : [...frameAcc, { id, data: [data] }];
    }, acc)
, []);

/**
 * Generates color props for chartes based on a given
 * "seed" color provided by the user
 */
export const generateTheme = (seedColor, data) => {
    const step = 360 / data.length;

    const dataColors = data.map((item, index) => {
        return {
            ...item,
            color: Color(seedColor).rotate(step * index).hex(),
        };
    });

    const contrastedColor = Color(seedColor).rotate(180).hex();

    return {
        highlightColor: contrastedColor,
        referenceLineColor: contrastedColor,
        referenceLabelColor: "#676D74",
        dataColors: dataColors,
    };
};

/**
 * Returns coordinates for drawing evenly spaced lines with a min and max offset
 */

export const getEvenLineCoords = (height, numLines = 5) => {
    //offsets are set to what recharts uses by default
    const minOffset = 20;
    const maxOffset = 5;

    const actualMin = 0 + minOffset;
    const actualMax = height - maxOffset;

    const range = actualMax - actualMin;

    const lines = Array.from( { length: numLines }, (v, i) => {
        return ( range / (numLines - 1) * (i) + minOffset);
    });

    return lines;
};
