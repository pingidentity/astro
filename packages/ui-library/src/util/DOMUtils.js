export const measureWidth = (elem) => {
    return elem.getBoundingClientRect().width;
};