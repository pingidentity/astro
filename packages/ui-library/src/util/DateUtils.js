/*eslint-disable valid-jsdoc*/

"use strict";
const moment = require("moment");
const dateFormat = "YYYY-MM-DD";
const timeFormat = "hh:MM:SSa";

const DateUtils = {
    /**
     * @alias module:util/Utils.formatDate
     * @desc Format the given timestamp in milliseconds as a YYYY-MM-DD date.
     *
     * @param {number} timestamp
     *     The timestamp (UNIX time * 1000)
     * @returns {string}
     *     Formatted date
     */
    formatDate: function (timestamp) {
        return moment.utc(timestamp).format(dateFormat);
    },

    /**
     * @alias module:util/Utils.formatTime
     * @desc Format the given timestamp in milliseconds as a YYYY-MM-DD hh:MM:SSa date.
     *
     * @param {number} timestamp
     *     The timestamp (UNIX time * 1000)
     * @returns {string}
     *     Formatted date+time
     */
    formatDateTime: function (timestamp) {
        return moment.utc(timestamp).format(`${dateFormat} ${timeFormat}`);
    },

    /**
     * @alias module:util/Utils.formatTime
     * @desc Format the given timestamp in milliseconds as a hh:MM:SSa time.
     *
     * @param {number} timestamp
     *     The timestamp (UNIX time * 1000)
     * @returns {string}
     *     Formatted time
     */
    formatTime: function (timestamp) {
        return moment.utc(timestamp).format(timeFormat);
    },
};

module.exports = DateUtils;