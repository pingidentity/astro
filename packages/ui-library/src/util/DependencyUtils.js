import React from "react";

/**
 * @module util/DependencyUtils
 * @desc Functions for showing nice errors when loading a component that has a requirement above our standard dependencies.
 */

/**
 * @alias module:util/DependencyUtils.usesStableContext
 *
 * @desc Throws an error if the React 16.3+ context API isn't available.
 *
 * @example
 *   usesStableContext()
 */

export const usesStableContext = (component) => {
    if (!React.createContext) {
        console.error(`${component} requires React 16.3`);
    }
};

/**
 * @alias module:util/DependencyUtils.usesFragments
 *
 * @desc Throws an error if React fragments aren't available.
 *
 * @example
 *   usesStableContext()
 */

export const usesFragments = (component) => {
    if (!React.Fragment) {
        console.error(`${component} requires React 16`);
    }
};
