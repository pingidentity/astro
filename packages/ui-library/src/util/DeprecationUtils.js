import React from "react";
import PropTypes from "prop-types";
import { isEqual } from "underscore";
import { isProduction } from "./Utils";

function getSubstituteMessage(sub) {
    return sub ? ` Use "${sub}" instead.` : "";
}

function getVersionMessage(version) {
    return version ? `version ${version}` : "a future release";
}

function renderMessage(message) {
    return message ? ` ${message}` : "";
}

const isDeprecatedValue = propValue =>
    ({ contains, value }) => contains ? (value.indexOf(propValue) > -1) : isEqual(value, propValue);

/**
 * @desc Fires prop deprecation warning with optional substitute and version to be removed.
 *       Use this as though it were a propType.
 *
 * @param {Object} propInfo - The deprecated prop info object.
 * @param {string} propInfo.message - An optional message tacked on to the end of the warning.
 * @param {string} propInfo.substitute - What to use instead of the deprecated prop.
 * @param {string} propInfo.version - The version number when the deprecation will occur.
 *
 * @param {enum} props - props of component
 * @param {string} propName - name of prop being validated
 *
 * @returns {{}} `Prop "${propName}" is deprecated and will be removed in a future release.`
 *
 * @example:
 *          class MyComponent extends React.Component {
 *              static propTypes = {
 *                  title: deprecatedProp({
 *                      message: "Here's how you move to 4.0.0",
 *                      substitute: "aDifferentProp",
 *                      version: "4.0.0"
 *                  })
 *              }
 *          }
 */
export const deprecatedProp = ({
    message,
    substitute,
    version
}) => (props, propName) => (props[propName] !== undefined)
    ? new Error(
        `Prop "${propName}" is deprecated and will be removed in ${getVersionMessage(version)}.` +
        `${substitute ? ` Use "${substitute}" instead.` : ""}${renderMessage(message)}`
    )
    : null;

/**
 * @desc Fires deprecation warnings for specific values of a prop with option substitutes
 *       and removal version. Use this as though it were a propType.
 *
 * @param {Object[]} propType - The regular propType validator to use. Required.
 * @param {Object[]} values - An array of deprecated value objects. These are checked for deep equality, not reference equality.
 * @param {string} value.message - An optional message tacked on to the end of the warning.
 * @param {*} values.value - The actual value being deprecated.
 * @param {*} values.substitute - The value to use instead.
 * @param {boolean} values.contains - Used for string values; if true, checks to see if value
 *      is contained in propValue instead of doing an equality check.
 * @param {string} version - The version number when the values will be deprecated. Optional.
 *
 * @param {enum} props - props of component
 * @param {string} propName - name of prop being validated
 *
 * @returns {{}} `Prop "${propName}" is deprecated and will be removed in a future release.`
 *
 * @example:
 *          class MyComponent extends React.Component {
 *              static propTypes = {
 *                  position: deprecatedPropValues({
 *                      propType: PropTypes.oneOf(["left", "top-left", "top-right"])
 *                      values: [
 *                          {
 *                              value: "top-left",
 *                              substitute: "left"
 *                          },
 *                          {
 *                              messages: "The top-right isn't part of new designs.",
 *                              value: "top-right"
 *                          }
 *                      ],
 *                      version: "4.0.0"
 *                  })
 *              }
 *          }
 */

export const deprecatedPropValues = ({
    customMessage,
    propType,
    values,
    version
}) => (props, propName, componentName) => {
    const { [propName]: propValue } = props;
    const deprecatedValue = values.find(isDeprecatedValue(propValue));
    // Default to regular propType validator if no values or if no deprecated values.
    if (!deprecatedValue) {
        return PropTypes.checkPropTypes({
            [propName]: propType
        }, props, propName, componentName);
    }

    const { message, substitute } = deprecatedValue;
    const versionMessage = getVersionMessage(version);
    return new Error(
        customMessage ||
        `The prop value "${propValue}" is deprecated and will be removed in ${versionMessage}.` +
            `${substitute ? ` Use "${substitute}" instead.` : ""}` +
            `${renderMessage(message)}`);
};

/**
 * @desc Higher order component that fires an error if it sees a deprecated prop or if it sees a deprecated value.
 *      Stops code from running when deprecated props are used.
 * @param {Object[]} props - An array of deprecated prop objects.
 * @param {string} props.message - An optional message tacked on to the end of a prop's warning. Used when deprecating whole props.
 * @param {string} props.name - The name of the deprecated prop.
 * @param {string} props.substitute - The substitute prop to use instead. Optional- used when deprecating whole props.
 * @param {Object[]} props.values - An array of deprecated value objects; used if deprecating specific values.
 * @param {string} props.values.message - An optional message tacked on to the end of a value's warning.
 * @param {*} props.values.value - The deprecated value.
 * @param {*} props.values.substitute - The substitute for the deprecated value. Optional.
 * @param {boolean} values.contains - Used for string values; if true, checks to see if value
 *      is contained in propValue instead of doing an equality check.
 *
 * @param {{}} WrappedComponent - The component to be wrapped in this HOC
 *
 * @returns {{}} This higher order component will either throw an error or return the wrapped component.
 *
 * @example: export default withDeprecatedProps([
 *  {
 *      name: "myProp",
 *      values: [
 *          {
 *              message: "Please do not use this awful thing.",
 *              name: "aValueYouShouldNotUse",
 *              substitute: "aMuchBetterValue"
 *          },
 *      ]
 *  },
 *  {
 *      message: "More info on this prop: (link)"
 *      name: "myDeprecatedProp",
 *      substitute: "aMuchBetterProp"
 *  }
 * ])(MyComponent)
 */
export const withDeprecatedProps = deprecated => WrappedComponent => props => {
    const invalidProps = deprecated.reduce((acc, { name, values = [], ...prop }) => {
        const { [name]: propValue } = props;
        if (propValue !== undefined) {
            const deprecatedValue = values.find(isDeprecatedValue(propValue));

            return [
                ...acc,
                {
                    ...prop,
                    ...deprecatedValue ? { deprecatedValue: { ...deprecatedValue, value: propValue } } : {},
                    name
                }
            ];
        } else {
            return acc;
        }
    }, []);

    if (invalidProps.length > 0) {
        const { displayName } = WrappedComponent;
        const propsMessages = invalidProps.reduce((propsMessage, {
            deprecatedValue: dv,
            message,
            name: propName,
            substitute: sub
        }) => {
            if (dv !== undefined) {
                return `${propsMessage}The prop value "${dv.value}" is deprecated and can no longer be used.${
                    getSubstituteMessage(dv.substitute)}${renderMessage(dv.message)}`;
            } else {
                const subMessage = sub ? `; use "${sub}" instead.` : ".";
                return `${propsMessage}Prop "${propName}" is deprecated and can no longer be used${
                    subMessage}${renderMessage(message)}\n`;
            }
        }, `Error in component "${displayName}":\n`);
        throw new Error(propsMessages.trim());
    } else {
        return <WrappedComponent {...props} />;
    }
};

/**
 * @desc Function that logs a warning that a component or module will be removed in a future release.
 * @param {Object} warning - The warning object.
 * @param {string} warning.message - An optional message tacked on to the end of the warning.
 * @param {string} warning.name - The name of the module or component being deprecated.
 * @param {string} warning.substitute - The component or module to be used instead.
 * @param {string} warning.type - The type of thing being deprecated. Default is "component".
 * @param {string} warning.version - The version when the module or component will be removed. Optional.
 *
 * @returns {} No return value.
 *
 * @example: deprecationWarning({
 *      message: "Here's some information about how to get ready for next version",
 *      name: "myRegrettableComponent",
 *      substitute: "myBetterComponent",
 *      version: "4.0.0"
 *  })
 */
export function deprecationWarning({
    message,
    name,
    type = "component",
    substitute,
    version
}) {
    console.error(
        `The ${type} "${name}" is deprecated and will be removed in ${
            getVersionMessage(version)
        }.${getSubstituteMessage(substitute)}${renderMessage(message)}`
    );
}

/**
 * @desc Function that logs a warning that a component or module has been removed. Causes a runtime error,
 *       which will prevent pages using the deprecated thing from loading and will fail related unit tests.
 * @param {Object} error - The error object.
 * @param {string} error.message - An optional message tacked on to the end of the warning.
 * @param {string} error.name - The name of the module or component being deprecated.
 * @param {string} error.substitute - The component or module to be used instead.
 * @param {string} error.type - The type of thing being deprecated. Default is "component".
 *
 * @returns {} No return value.
 *
 * @example: deprecationError({
 *      message: "Here's an explanation of the substitute value.",
 *      name: "myRegrettableComponent",
 *      substitute: "myBetterComponent"
 *  })
 */
export function deprecationError({
    message,
    name,
    type = "component",
    substitute
}) {
    throw new Error(
        `The ${type} "${name}" is deprecated and has been removed.${
            getSubstituteMessage(substitute)}${renderMessage(message)}`
    );
}

/**
 * @desc Function specifically for the v4 Cannonball release. Shows a warning about changing behavior
 * @param {string} message - The message that will appear between the two pieces of boilerplate.
 *
 * @returns {} No return value.
 *
 * @example: cannonballChangeWarning({
 *      message: "This component will act much differently. Add 'much-differently' flag to preview.",
 *  })
 */
export function cannonballChangeWarning({
    message
}) {
    if (!isProduction()) {
        console.warn(
            `WARNING potential 4.0.0 breaking change: ` +
            message +
            ` For more information: https://uilibrary.ping-eng.com/3.42.0-SNAPSHOT/index.html#/?selectedSection=V4BreakingChanges&selectedNode=V4BreakingChanges&root=Documentation`
        );
    }
}

/**
 * @desc Function specifically for the v4 Cannonball release. Shows a warning about this specific behavior change
 * @param {string} name - The the name of the component
 *
 * @returns {} No return value.
 *
 * @example: cannonballProgressivelyStatefulWarning({
 *      name: "FormTextArea",
 *  })
 */
export function cannonballProgressivelyStatefulWarning({
    name
}) {
    cannonballChangeWarning({
        message: `The stateless prop will be removed from ${name}. ` +
            `Instead of distinguishing between stateless and stateful versions, ` +
            `the component will instead control the state of a subset of its props when those ` +
            `have not been defined.`
    });
}

/**
 * @desc Function specifically for the v4 Cannonball release. Shows a warning about portal behavior
 * @param {string} name - The name of the component
 *
 * @returns {} No return value.
 *
 * @example: cannonballPortalWarning({
 *      name: "Calendar",
 *  })
 */
export function cannonballPortalWarning({
    name
}) {
    cannonballChangeWarning({
        message: `${name} will render using a portal and auto-positioning.`
    });
}
