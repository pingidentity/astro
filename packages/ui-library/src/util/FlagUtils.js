import PropTypes from "prop-types";

export const allFlags = [
    "pass-value",
    // v4 flags
    "p-stateful",
    "use-portal",
    "add-button-margin",
    "expandable-row-class",
    "fix-discard-button",
    "fixed-messages-constants",
    "true-default",
    "v4",
    "new-alert-modal",
];

export const flagsPropType = PropTypes.arrayOf(PropTypes.oneOf(allFlags));

export const getFlags = ({ props, context }) => props.flags || (context && context.flags) || [];

export const hasFlag = (component, flag) => {
    const flags = getFlags(component);
    return flags.includes(flag);
};
