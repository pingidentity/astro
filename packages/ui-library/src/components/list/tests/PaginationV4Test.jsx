window.__DEV__ = true;

import { mount } from "enzyme";
import StateContainer from "../../utils/StateContainer";
import { allFlags } from "../../../util/FlagUtils";

jest.dontMock("../Pagination");
jest.dontMock("../../rows/expandable-row/ExpandableRow");

describe("Pagination", function () {

    var React = require("react"),
        ReactTestUtils = require("react-dom/test-utils"),
        TestUtils = require("../../../testutil/TestUtils"),
        Utils = require("../../../util/Utils"),
        Pagination = require("../Pagination"),
        ExpandableRow = require("../../rows/expandable-row/ExpandableRow"),
        callback,
        component,
        pageLinks,
        top,
        topLinks;

    beforeEach(function () {
        callback = jest.fn();
        component = ReactTestUtils.renderIntoDocument(
            <Pagination
                data-id="test-pagination"
                page={2}
                perPage = {5}
                total = {100}
                onValueChange={callback}>
                <ExpandableRow className="row" key={1} />
                <ExpandableRow className="row" key={2} />
                <ExpandableRow className="row" key={3} />
                <ExpandableRow className="row" key={4} />
                <ExpandableRow className="row" key={5} />
            </Pagination>
        );

        pageLinks = TestUtils.scryRenderedDOMNodesWithClass(component, "page-links");
        top = pageLinks[0];
        topLinks = top.childNodes;

    });

    it ("renders short pages navigation", function () {

        component = ReactTestUtils.renderIntoDocument(
            <Pagination
                perPage = {5}
                total = {30}
                onValueChange={callback}
                flags={allFlags}
            >
                <ExpandableRow className="row" key={1} />
                <ExpandableRow className="row" key={2} />
                <ExpandableRow className="row" key={3} />
                <ExpandableRow className="row" key={4} />
                <ExpandableRow className="row" key={5} />
            </Pagination>
        );

        pageLinks = TestUtils.scryRenderedDOMNodesWithClass(component, "page-links");
        top = pageLinks[0];
        topLinks = top.childNodes;

        // expect 2 lists of page links
        expect(pageLinks.length).toBe(2);

        // expect 8 child dom nodes in page links, no ellipsis
        expect(topLinks.length).toEqual(8);

        // expect the first child to be "<<"
        expect(topLinks[0].childNodes[0].className).toBe("icon-previous");

        // expect the highest page to be 6
        expect(topLinks[6].textContent).toBe("6");

    });

    it ("renders expanded pages navigation", function () {

        // expect 2 lists of page links
        expect(pageLinks.length).toBe(2);
        // expect 10 child dom nodes in page links, 2 for next and previous, 2 for first and last pages, one for ellipsis and 5 for pages
        expect(topLinks.length).toEqual(10);

        // expect the first child to be "<<"
        expect(topLinks[0].childNodes[0].className).toBe("icon-previous");

        // expect the highest page to be 20
        expect(topLinks[8].textContent).toBe("20");

    });

    it ("renders no pages navigation", function () {

        component = ReactTestUtils.renderIntoDocument(
            <Pagination
                perPage = {5}
                total = {4}
                onValueChange={callback}
                flags={allFlags}
            >
                <ExpandableRow className="row" key={1} />
                <ExpandableRow className="row" key={2} />
                <ExpandableRow className="row" key={3} />
                <ExpandableRow className="row" key={4} />
                <ExpandableRow className="row" key={5} />
            </Pagination>
        );

        pageLinks = TestUtils.scryRenderedDOMNodesWithClass(component, "page-links");

        topLinks = pageLinks[0].childNodes;
        var bottomLinks = pageLinks[1].childNodes;

        // expect 2 lists of page links
        expect(pageLinks.length).toBe(2);

        // expect no child DOM nodes
        expect(topLinks.length).toEqual(0);
        expect(bottomLinks.length).toEqual(0);
    });

    it ("renders navigation using totalPages instead of total and perPage", function () {

        component = ReactTestUtils.renderIntoDocument(
            <Pagination
                totalPages={3}
                perPage={3}
                onValueChange={callback}
                flags={allFlags}
            >
                <ExpandableRow className="row" key={1} />
                <ExpandableRow className="row" key={2} />
                <ExpandableRow className="row" key={3} />
            </Pagination>
        );

        pageLinks = TestUtils.scryRenderedDOMNodesWithClass(component, "page-links");
        topLinks = pageLinks[0].childNodes;

        // expect 2 lists of page links
        expect(pageLinks.length).toBe(2);

        // expect the first child to be "<<"
        expect(topLinks[0].childNodes[0].className).toBe("icon-previous");

        // expect numeric links to be the correct number
        expect(topLinks[1].textContent).toBe("1");
        expect(topLinks[2].textContent).toBe("2");
        expect(topLinks[3].textContent).toBe("3");

        // expect the last child to be ">>"
        expect(topLinks[4].childNodes[0].className).toBe("icon-next");
    });

    it ("trigger page change callback", function () {
        ReactTestUtils.Simulate.click(topLinks[1]);
        expect(callback).toBeCalledWith({ first: 0, last: 5, page: 1 });
    });

    it ("is not triggering callback for same page", function () {
        ReactTestUtils.Simulate.click(topLinks[0]);
        expect(callback).not.toBeCalled();
    });

    it ("renders children content", function () {
        var children = TestUtils.scryRenderedDOMNodesWithClass(component, "row");
        expect(children.length).toEqual(5);
    });

    it ("verify default data-id", function () {
        var paging = ReactTestUtils.renderIntoDocument(
            <Pagination
                totalPages={3}
                perPage={3}
                onValueChange={callback}
                flags={allFlags}
            >
                <ExpandableRow className="row" key={1} />
            </Pagination>
        );
        var test = TestUtils.findRenderedDOMNodeWithDataId(paging, "pagination");
        expect(test).toBeTruthy();
    });

    it ("verify large number of pages", function () {
        component = ReactTestUtils.renderIntoDocument(
            <Pagination
                stateless={true}
                data-id="test-pagination"
                page={6}
                perPage = {5}
                total = {100}
                onValueChange={callback}
                flags={allFlags}
            >
                <ExpandableRow className="row" key={1} />
                <ExpandableRow className="row" key={2} />
                <ExpandableRow className="row" key={3} />
                <ExpandableRow className="row" key={4} />
                <ExpandableRow className="row" key={5} />
            </Pagination>
        );
        pageLinks = TestUtils.scryRenderedDOMNodesWithClass(component, "page-links");
        top = pageLinks[0];
        topLinks = top.childNodes;

        ReactTestUtils.Simulate.click(topLinks[9]);
        expect(callback).toBeCalledWith({ first: 95, last: 100, page: 20 });
    });

    it ("verify no pages", function () {
        component = ReactTestUtils.renderIntoDocument(
            <Pagination
                stateless={true}
                data-id="test-pagination"
                page={1}
                perPage = {5}
                total = {1}
                onValueChange={callback}
                flags={allFlags}
            >
                <ExpandableRow className="row" key={1} />
                <ExpandableRow className="row" key={2} />
                <ExpandableRow className="row" key={3} />
                <ExpandableRow className="row" key={4} />
                <ExpandableRow className="row" key={5} />
            </Pagination>
        );
        pageLinks = TestUtils.scryRenderedDOMNodesWithClass(component, "page-links");
        top = pageLinks[0];
        topLinks = top.childNodes;
        expect(topLinks.length).toEqual(0);
    });

    it("throws error when deprecated prop 'id' is passed in", function () {
        var expectedError = new Error(Utils.deprecatePropError("id", "data-id"));

        expect(function () {
            ReactTestUtils.renderIntoDocument(
                <Pagination id="foo" totalPages={2} perPage={1} onValueChange={jest.fn()} flags={allFlags}>
                    <ExpandableRow className="row" key={1} />
                    <ExpandableRow className="row" key={2} />
                </Pagination>
            );
        }).toThrow(expectedError);
    });

    it("throws error when deprecated prop 'controlled' is passed in", function () {
        var expectedError = new Error(Utils.deprecatePropError("controlled", "stateless"));

        expect(function () {
            ReactTestUtils.renderIntoDocument(
                <Pagination controlled={true} totalPages={2} perPage={1} onValueChange={jest.fn()}>
                    <ExpandableRow className="row" key={1} />
                    <ExpandableRow className="row" key={2} />
                </Pagination>
            );
        }).toThrow(expectedError);
    });

    it("throws error when deprecated prop 'onChange' is passed in", function () {
        var expectedError = new Error(Utils.deprecatePropError("onChange", "onValueChange"));

        expect(function () {
            ReactTestUtils.renderIntoDocument(
                <Pagination onChange={jest.fn()} totalPages={2} perPage={1} flags={allFlags}>
                    <ExpandableRow className="row" key={1} />
                    <ExpandableRow className="row" key={2} />
                </Pagination>
            );
        }).toThrow(expectedError);
    });

    it("does not throw error when deprecated prop 'id' is passed in in production", function () {
        var expectedError = new Error(Utils.deprecatePropError("id", "data-id"));
        process.env.NODE_ENV = "production";

        expect(function () {
            ReactTestUtils.renderIntoDocument(
                <Pagination id="foo" totalPages={2} perPage={1} onValueChange={jest.fn()} flags={allFlags}>
                    <ExpandableRow className="row" key={1} />
                    <ExpandableRow className="row" key={2} />
                </Pagination>
            );
        }).not.toThrow(expectedError);

        process.env.NODE_ENV = "";
    });

    // Progressively stateful tests
    it("renders a progressively stateful component when the p-stateful flag is passed in", () => {
        const pagination = mount(
            <Pagination
                flags={allFlags}
                totalPages={1}
            />
        );

        expect(pagination.find(StateContainer).exists()).toBeTruthy();
    });

    it("updates page in p-stateful version", () => {
        const pagination = mount(
            <Pagination
                flags={allFlags}
                totalPages={10}
            />
        );

        expect(pagination.find("[data-id=\"bottomPageLinksfirst\"]").hasClass("active")).toBeTruthy();

        pagination.find("[data-id=\"bottomPageLinks2\"]").simulate("click");

        expect(pagination.find("[data-id=\"bottomPageLinksfirst\"]").hasClass("active")).toBeFalsy();
        expect(pagination.find("[data-id=\"bottomPageLinks2\"]").hasClass("active")).toBeTruthy();
    });

});
