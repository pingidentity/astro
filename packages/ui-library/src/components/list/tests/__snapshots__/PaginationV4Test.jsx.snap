// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination data-id's don't change 1`] = `
Array [
  "test-pagination",
  "topPageLinks",
  "topPageLinksprevious",
  "topPageLinksfirst",
  "topPageLinks2",
  "topPageLinks3",
  "topPageLinks4",
  "topPageLinks5",
  "topPageLinks6",
  "topPageLinks20",
  "topPageLinksnext",
  "expandable-row",
  "expand-btn",
  "edit-btn",
  "delete-btn",
  "expandable-row",
  "expand-btn",
  "edit-btn",
  "delete-btn",
  "expandable-row",
  "expand-btn",
  "edit-btn",
  "delete-btn",
  "expandable-row",
  "expand-btn",
  "edit-btn",
  "delete-btn",
  "expandable-row",
  "expand-btn",
  "edit-btn",
  "delete-btn",
  "expandable-row",
  "expand-btn",
  "edit-btn",
  "delete-btn",
  "expandable-row",
  "expand-btn",
  "edit-btn",
  "delete-btn",
  "bottomPageLinks",
  "bottomPageLinksprevious",
  "bottomPageLinksfirst",
  "bottomPageLinks2",
  "bottomPageLinks3",
  "bottomPageLinks4",
  "bottomPageLinks5",
  "bottomPageLinks6",
  "bottomPageLinks20",
  "bottomPageLinksnext",
]
`;
